python/mead/downloader.py

```diff
@@ -0,0 +1,287 @@
+
+    except:  # this is too broad but there are too many exceptions to handle separately
+        raise RuntimeError("failed to download data from [url]: {} [to]: {}".format(url, path_to_save))
+    return path_to_save
+
+
+def validate_url(url):
+    regex = re.compile(
+        r'^(?:http|ftp)s?://'  # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+        r'localhost|'  # localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+        r'(?::\d+)?'  # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    return re.match(regex, url) is not None
+
+
+def update_cache(key, data_download_cache):
+    dcache = read_json(os.path.join(data_download_cache, DATA_CACHE_CONF))
+    if key not in dcache:
+        return
+    del dcache[key]
+    write_to_json(dcache, os.path.join(data_download_cache, DATA_CACHE_CONF))
+
+
+def check_sanity_file(file_loc, data_dcache=None, key=None):
+    if os.path.exists(file_loc) and os.path.isfile(file_loc) and not mime_type(file_loc) == "text/html":
+        # dropbox doesn't give 404 in case the file does not exist, produces an HTML. The actual files are never HTMLs.
+        return True
+    else:
+        delete_old_copy(file_loc)
+        if key is not None:  # cache file validation
+            update_cache(key, data_dcache)
+        return False
+
+
+def check_sanity_dir(dir_loc, dataset_desc, data_dcache=None, ignore_file_check=False, key=None):
+    if not os.path.exists(dir_loc) or not os.path.isdir(dir_loc):
+        update_cache(key, data_dcache)
+        return False
+    if ignore_file_check:  # for enc_dec tasks there's no direct downloads
+        return True
+    files = [os.path.join(dir_loc, dataset_desc[k]) for k in dataset_desc if k.endswith("_file")]
+    for file in files:
+        if not check_sanity_file(file, key, data_dcache):
+            return False
+    return True
+
+
+def read_json(filepath):
+    if not os.path.exists(filepath):
+        return {}
+    with open(filepath) as f:
+        j_con = json.load(f)
+    return j_con
+
+
+def write_to_json(content, filepath):
+    with open(filepath, "w") as f:
+        json.dump(content, f, indent=True)
+
+
+class Downloader(object):
+    def __init__(self, data_download_cache, cache_ignore):
+        super(Downloader, self).__init__()
+        self.cache_ignore = cache_ignore
+        self.data_download_cache = data_download_cache
+
+    def download(self):
+        pass
+
+
+class SingleFileDownloader(Downloader):
+    def __init__(self, dataset_file, data_download_cache, cache_ignore=False):
+        super(SingleFileDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_file = dataset_file
+        self.data_download_cache = data_download_cache
+        self.zipd = {'application/gzip': extract_gzip, 'application/zip': extract_zip}
```

Comment:
Appears to be shared between all impls.  Push up.  Also seems that it could be defined as a static